Fixes #4335: Addition of support for uploading performance metric logs

Sarthak2601 opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #4335
- Introduces support for uploading performance metric logs to the remote service. 
- The PR modifies files related to work manager to add and queue work request for uploading metric logs. It also modifies files related to unpacking and repacking of event logs into bundles for logging to firebase. 
- Added tests to verify these functionalities as well. 

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

Sarthak2601 committed:
qualifiers and constants for metric record and upload times

Sarthak2601 committed:
comments

Sarthak2601 committed:
dagger provides for flags

Sarthak2601 committed:
rename to enablePerformanceMetricCollection

Sarthak2601 committed:
initial proto

Sarthak2601 committed:
nit fixes

Sarthak2601 committed:
Merge branch 'performance-metrics-platform-params' into performance-metrics-proto

Sarthak2601 committed:
nit fixes

Sarthak2601 committed:
comments.

Sarthak2601 committed:
nits

Sarthak2601 committed:
nit

Sarthak2601 committed:
Merge branch 'performance-metrics-platform-params' into performance-metrics-proto

Sarthak2601 committed:
updates.

Sarthak2601 committed:
updates.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-platform-params

Sarthak2601 committed:
Merge branch 'performance-metrics-platform-params' into performance-metrics-proto

Sarthak2601 committed:
metric log inclusion.

Sarthak2601 committed:
base setup.

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-platform-params

Sarthak2601 committed:
name correction.

Sarthak2601 committed:
Merge branch 'performance-metrics-platform-params' into performance-metrics-proto

Sarthak2601 committed:
nits.

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

Sarthak2601 committed:
nits.

Sarthak2601 committed:
storage comment

Sarthak2601 committed:
pss comment

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-proto

Sarthak2601 committed:
network usage comment.

Sarthak2601 committed:
network usage comment - part 2.

Sarthak2601 committed:
metric addition in proto definitions.

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

Sarthak2601 committed:
metricLog --> loggableMetric

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

Sarthak2601 committed:
proto definitions in oppiaLogger

Sarthak2601 committed:
wording update for transmission

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

Sarthak2601 committed:
dependency/api updates.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
metric controller and oppiaLogger

Sarthak2601 committed:
nits

Sarthak2601 committed:
logger-controller pattern, single utils

Sarthak2601 committed:
nits

Sarthak2601 committed:
comments

Sarthak2601 committed:
log uploading support

Sarthak2601 committed:
nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
performance metric event logger nits.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

# Conflicts:
#	utility/src/main/java/org/oppia/android/util/logging/performancemetrics/PerformanceMetricsEventLogger.kt

Sarthak2601 committed:
logger reinstated.

Sarthak2601 committed:
nits

Sarthak2601 committed:
deleting unused interfaces.

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-proto

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

Sarthak2601 committed:
log generator to metric log scheduler

Sarthak2601 committed:
nits

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkManagerInitializer.kt

BenHenning commented on domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkManagerInitializer.kt:
```diff
@@ -118,6 +131,10 @@ class LogUploadWorkManagerInitializer @Inject constructor(
     val workManager = WorkManager.getInstance(context)
     logUploader.enqueueWorkRequestForEvents(workManager, workRequestForUploadingEvents)
     logUploader.enqueueWorkRequestForExceptions(workManager, workRequestForUploadingExceptions)
+    logUploader.enqueueWorkRequestForPerformanceMetrics(
```
Is there a test to verify this is working correctly?

Sarthak2601 replied:
added.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -97,12 +103,30 @@ class LogUploadWorker private constructor(
+  private suspend fun uploadPerformanceMetrics(): Result {
+    return try {
+      val performanceMetricsLogs = performanceMetricsController.getMetricLogStoreList()
+      performanceMetricsLogs.let {
```
What purpose does using 'let' serve here? It might be decreasing the readability of the code since it seems unnecessary.

Sarthak2601 replied:
replaced with forEach to enhance readability.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt:
```diff
@@ -18,4 +19,11 @@ class LogReportingModule {
+
+  @Provides
+  @Singleton
+  fun providePerformanceMetricsEventLogger(
```
Any test to verify this binding?

Sarthak2601 replied:
I've added tests to verify this binding with the help of a fake in the earlier PR. Does that suffice?

BenHenning replied:
I suggest adding an explicit LogReportingModuleTest (see other module tests for an idea)--they're generally quick to setup, and are useful for directly verifying the exact binding.

Sarthak2601 replied:
done.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -47,6 +52,7 @@ class LogUploadWorker private constructor(
       when (inputData.getStringFromData(WORKER_CASE_KEY)) {
         EVENT_WORKER -> uploadEvents()
         EXCEPTION_WORKER -> uploadExceptions()
+        PERFORMANCE_METRICS_WORKER -> uploadPerformanceMetrics()
```
Any tests to verify that this is working as expected?

Sarthak2601 replied:
added.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -84,6 +105,43 @@ class EventBundleCreator @Inject constructor(
+    }?.metricName ?: "unknown_loggable_metric"
+  }
+
+  private fun OppiaMetricLog.LoggableMetric.convertToLoggableMetricType():
```
A nit on member ordering: since the method above appears after the other public API method in this class, I feel like this helper should appear after its helper (since it appears next in the class order).

Sarthak2601 replied:
done.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -84,6 +105,43 @@ class EventBundleCreator @Inject constructor(
+    bundle.putString("network_type", oppiaMetricLog.networkType.toAnalyticsName())
+    bundle.putString("current_screen", oppiaMetricLog.currentScreen.toAnalyticsName())
+    return oppiaMetricLog.loggableMetric.convertToLoggableMetricType()?.also { loggableMetric ->
+      // No allowance for user IDs to be logged since learner study feature is disabled for
+      // performance metrics logging as of now.
```
Can probably just be: ``No performance metrics need to be tied to user IDs.`` since I don't think we have long-term plans for this, either.

Sarthak2601 replied:
done.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -84,6 +105,43 @@ class EventBundleCreator @Inject constructor(
     }?.activityName ?: "unknown_activity_context"
   }
 
+  fun fillPerformanceMetricsEventBundle(oppiaMetricLog: OppiaMetricLog, bundle: Bundle): String {
```
Please add a KDoc for this method since it's an addition to the class's public API.

Sarthak2601 replied:
done.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -84,6 +105,43 @@ class EventBundleCreator @Inject constructor(
     }?.activityName ?: "unknown_activity_context"
   }
 
+  fun fillPerformanceMetricsEventBundle(oppiaMetricLog: OppiaMetricLog, bundle: Bundle): String {
```
Any tests for this new functionality?

Sarthak2601 replied:
added.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt:
```diff
@@ -19,7 +21,7 @@ class FirebaseEventLogger private constructor(
   private val networkConnectionUtil: NetworkConnectionUtil,
   private val eventBundleCreator: EventBundleCreator
-) : EventLogger {
+) : EventLogger, PerformanceMetricsEventLogger {
```
How painful would it be to rename the existing event logger? I'm just a bit concerned that it's too general sounding now that we're specializing performance metrics.

Sarthak2601 replied:
We'll need to do the following: 
- changes in all the production files where eventLogger is used,
- changes in test files. 
- changes in dagger modules. 
- refactoring of fakeEventlogger.

It might not be too painful -- we can file this issue as a good-second-issue.

Sarthak2601 replied:
Filed an issue for this -- https://github.com/oppia/oppia-android/issues/4405

BenHenning replied:
That seems reasonable--thanks for filing the issue!

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt:
```diff
@@ -60,5 +80,8 @@ class FirebaseEventLogger private constructor(
     fun create(): EventLogger =
       FirebaseEventLogger(firebaseAnalytics, networkConnectionUtil, eventBundleCreator)
+
+    fun createPerformanceMetricEventLogger(): PerformanceMetricsEventLogger =
```
Please add a KDoc for this. Also, any new tests for this method?

Sarthak2601 replied:
added kdoc.

BenHenning reviewed (commented):
Thanks @Sarthak2601. Took a first pass--PTAL.

Sarthak2601 committed:
nits -- kdoc, code placing, readability

Sarthak2601 committed:
nits -- kdoc, code placing, readability

Sarthak2601 committed:
worker functions to schedule logging, renaming

Sarthak2601 committed:
nits.

Sarthak2601 committed:
clearing up build errors.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

# Conflicts:
#	utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt

Sarthak2601 committed:
fixing build errors.

Sarthak2601 committed:
removal of dependency from oppiaLogger and basing the entire implementation on Logger-Controller

Sarthak2601 committed:
performanceMetricController tests + fakeLogger

Sarthak2601 committed:
fakeLogger tests

Sarthak2601 committed:
nits

Sarthak2601 committed:
tests for the new logger.

Sarthak2601 committed:
tests for the worker + renaming of package to logScheduler

Sarthak2601 committed:
tests for work manager initializer

Sarthak2601 committed:
initial base setup for utils test

Sarthak2601 committed:
nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkManagerInitializer.kt

Sarthak2601 committed:
nit

Sarthak2601 committed:
worker and initializer tests

Sarthak2601 committed:
event bundle creator tests

BenHenning commented on utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt:
```diff
@@ -133,6 +176,118 @@ class EventBundleCreatorTest {
   }
 
+  @Test
+  fun testFillMetricsBundle_eventWithDefaultedContext_fillsDetailsAndRetsUnknownContext() {
```
``eventWithDefaultloggableMetric``? There's no "context" field name in this proto.

Sarthak2601 replied:
done.

BenHenning reviewed (commented):
Thanks @Sarthak2601! Just had a few comments--PTAL.

Sarthak2601 committed:
module tests

Sarthak2601 committed:
nits

Sarthak2601 commented:
@BenHenning if this PR looks good to you, please don't merge after approval -- we'll merge the stacked PRs one by one otherwise all file changes will get stuffed up in one PR which will then become very tedious to review and manage :)

BenHenning commented on utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt:
```diff
@@ -177,7 +177,7 @@ class EventBundleCreatorTest {
 
   @Test
-  fun testFillMetricsBundle_eventWithDefaultedContext_fillsDetailsAndRetsUnknownContext() {
+  fun testFillMetricsBundle_eventWithDefaultLoggableMetric_fillsDetailsAndRetsUnknownContext() {
```
```suggestion
  fun testFillMetricsBundle_eventWithDefaultLoggableMetric_fillsDetailsAndRetsUnknownLog() {
```

Maybe?

Sarthak2601 replied:
done.

BenHenning commented on utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt:
```diff
@@ -0,0 +1,150 @@
+
+  @Test
+  fun testModule_injectsProductionImplementationOfEventLogger() {
+    Truth.assertThat(eventLogger).isInstanceOf(FirebaseEventLogger::class.java)
```
Nit: please directly import assertThat for parity with other test suites.

Sarthak2601 replied:
done.

BenHenning commented on utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  @Test
+  fun testModule_injectsProductionImplementationOfEventLogger() {
```
Thanks for adding a test for the existing binding as well!

BenHenning reviewed (commented):
Thanks @Sarthak2601. Just 2 nits, otherwise LGTM.

BenHenning commented:
> @BenHenning if this PR looks good to you, please don't merge after approval -- we'll merge the stacked PRs one by one otherwise all file changes will get stuffed up in one PR which will then become very tedious to review and manage :)

I agree--I generally only merge branches that go into ``develop``.

Also, it looks like a lot of CI checks are failing, so I slightly withdraw my previous "LGTM" until these are also addressed so that the PR is green.

Sarthak2601 committed:
performanceMetricsUtils tests

Sarthak2601 committed:
nits

Sarthak2601 committed:
renamed logUpload to logReport + nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

# Conflicts:
#	domain/src/test/java/org/oppia/android/domain/oppialogger/loguploader/LogReportWorkManagerInitializerTest.kt

Sarthak2601 committed:
review updates.

Sarthak2601 committed:
nit

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-log-creation

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-proto

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

Sarthak2601 committed:
review updates - 2

Sarthak2601 committed:
isAppInForeground revamp

Sarthak2601 committed:
isAppInForeground revamp

Sarthak2601 committed:
nits

Sarthak2601 committed:
domain tests repair

Sarthak2601 committed:
app tests repair

Sarthak2601 committed:
nits.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
nit

Sarthak2601 committed:
test update.

Sarthak2601 committed:
test update.

Sarthak2601 committed:
nit

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

Sarthak2601 committed:
todo issue number update -- static check

Sarthak2601 committed:
comments.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
utils test + nits

Sarthak2601 committed:
utils test + nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
test rearrangement

Sarthak2601 committed:
nits

Sarthak2601 committed:
test fixes.

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-proto

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

# Conflicts:
#	app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivityTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsFragmentTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/learneranalytics/ProfileAndDeviceIdFragmentTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/options/OptionsFragmentTest.kt
#	app/src/test/java/org/oppia/android/app/testing/options/OptionsFragmentTest.kt

Sarthak2601 committed:
nits

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogReportWorkManagerInitializer.kt
#	utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt

Sarthak2601 committed:
nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
memory and storage tier updates

Sarthak2601 committed:
Merge branch 'performance-metrics-proto' into performance-metrics-log-creation

Sarthak2601 committed:
memory and storage tier updates

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-log-creation

# Conflicts:
#	model/src/main/proto/performance_metrics.proto

Sarthak2601 committed:
updates.

Sarthak2601 committed:
nits

Sarthak2601 committed:
test file exemptions refactor due to file renaming.

Sarthak2601 committed:
additional tests, nit fixes.

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-log-creation

Sarthak2601 committed:
activity manager shadow and assessor test

Sarthak2601 committed:
nit

Sarthak2601 committed:
nit

Sarthak2601 committed:
shadow traffic stats + assessor test fix.

Sarthak2601 committed:
custom shadow tests

Sarthak2601 committed:
module deps

Sarthak2601 committed:
tests

Sarthak2601 committed:
nits

Sarthak2601 committed:
metricLogScheduler refactor and test

Sarthak2601 committed:
exemption removal

Sarthak2601 committed:
nits

Sarthak2601 committed:
updates.

Sarthak2601 committed:
variable nits + parameterized test exemption

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-log-creation

# Conflicts:
#	app/src/main/java/org/oppia/android/app/application/ApplicationComponent.kt
#	scripts/assets/test_file_exemptions.textproto

Sarthak2601 committed:
app component dependencies.

Sarthak2601 committed:
nits

Sarthak2601 committed:
logging module bazel build, config module creation

Sarthak2601 committed:
testing robolectric bazel module update

Sarthak2601 committed:
domain bazel build fixes

Sarthak2601 committed:
working oppia bazel build

Sarthak2601 committed:
nits.

Sarthak2601 committed:
nits.

Sarthak2601 committed:
nit for bazel building

Sarthak2601 committed:
addition of test file for module

Sarthak2601 committed:
bazel tests fixes

Sarthak2601 committed:
more fixes.

Sarthak2601 committed:
reformatting

Sarthak2601 committed:
nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-log-creation

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

Sarthak2601 committed:
dep fix

Sarthak2601 committed:
deps fix

Sarthak2601 committed:
deps fix

Sarthak2601 committed:
nit

Sarthak2601 committed:
nits

Sarthak2601 committed:
nits

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

Sarthak2601 committed:
gradle test fix.

Sarthak2601 committed:
bazel oppia build

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-log-creation

# Conflicts:
#	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt

Sarthak2601 commented:
> > @BenHenning if this PR looks good to you, please don't merge after approval -- we'll merge the stacked PRs one by one otherwise all file changes will get stuffed up in one PR which will then become very tedious to review and manage :)
> 
> I agree--I generally only merge branches that go into `develop`.
> 
> Also, it looks like a lot of CI checks are failing, so I slightly withdraw my previous "LGTM" until these are also addressed so that the PR is green.

@BenHenning I've fixed all CI tests, PTAL -- thanks!

Sarthak2601 committed:
test fixes -- renaming.

Sarthak2601 committed:
deps addition

Sarthak2601 committed:
previous merge correction

BenHenning reviewed (commented):
Thanks @Sarthak2601! The PR generally LGTM. Will wait to approve after the base PR is merged and this one is updated to be based on develop.

Sarthak2601 committed:
nits

Sarthak2601 committed:
updates.

Sarthak2601 committed:
nits

Sarthak2601 committed:
bazel update.

Sarthak2601 committed:
nit

Sarthak2601 committed:
nits.

Sarthak2601 committed:
Merge branch 'develop' into performance-metrics-log-creation

Sarthak2601 committed:
Merge branch 'performance-metrics-log-creation' into performance-metric-log-upload

Sarthak2601 committed:
Merge branch 'develop' into performance-metric-log-upload

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogReportWorkManagerInitializer.kt
#	domain/src/test/java/org/oppia/android/domain/oppialogger/loguploader/LogReportWorkManagerInitializerTest.kt
#	utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt

BenHenning reviewed (approved):
Thanks @Sarthak2601! LGTM. Feel free to merge whenever you need to.

BenHenning commented:
(Unassigning Rajat since this PR doesn't need his review).

Sarthak2601 committed:
Merge branch 'develop' into performance-metric-log-upload
